src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,702 @@
+    Field(Type(HeapType::ext, NonNullable), Mutable),
+    Field(Type(HeapType::none, NonNullable), Mutable),
+    Field(Type(HeapType::nofunc, NonNullable), Mutable),
+    Field(Type(HeapType::noext, NonNullable), Mutable),
+  }};
+}
+
+struct MinimizeRecGroups : Pass {
+  // The types we are optimizing and their indices in this list.
+  std::vector<HeapType> types;
+
+  // A global ordering on all types, including public types. Used to define a
+  // total ordering on rec group shapes.
+  std::unordered_map<HeapType, size_t> typeIndices;
+
+  // As we process strongly connected components, we will construct output
+  // recursion groups here.
+  std::vector<RecGroupInfo> groups;
+
+  // For each shape of rec group we have created, its index in `groups`.
+  std::unordered_map<RecGroupShape, size_t> groupShapeIndices;
+
+  // When we find that two groups are isomorphic to (i.e. permutations of)
+  // each other, we combine their equivalence classes and choose a new class
+  // representative to use to disambiguate the groups.
+  DisjointSets equivalenceClasses;
+
+  void run(Module* module) override {
+    // There are no recursion groups to minimize if GC is not enabled.
+    if (!module->features.hasGC()) {
+      return;
+    }
+
+    initBrandOptions();
+
+    types = ModuleUtils::getPrivateHeapTypes(*module);
+    for (auto type : ModuleUtils::collectHeapTypes(*module)) {
+      typeIndices.insert({type, typeIndices.size()});
+    }
+
+    // The number of types to optimize is an upper bound on the number of
+    // recursion groups we will emit.
+    groups.reserve(types.size());
+
+    // Compute the strongly connected components and ensure they form distinct
+    // recursion groups.
+    for (auto scc : TypeSCCs(types)) {
+      [[maybe_unused]] size_t index = equivalenceClasses.addSet();
+      assert(index == groups.size());
+      groups.emplace_back();
+
+      // The SCC is not necessarily topologically sorted to have the supertypes
+      // come first. Fix that.
+      std::vector<HeapType> sccTypes(scc.begin(), scc.end());
+      auto deps = createSubtypeGraph(sccTypes);
+      auto permutation = *TopologicalOrders(deps).begin();
+      groups.back().group.resize(sccTypes.size());
+      for (size_t i = 0; i < sccTypes.size(); ++i) {
+        groups.back().group[i] = sccTypes[permutation[i]];
+      }
+      updateShape(index);
+    }
+
+    rewriteTypes(*module);
+  }
+
+  void initBrandOptions() {
+    // Initialize the field options for brand types lazily here to avoid
+    // depending on global constructor ordering.
+    [[maybe_unused]] static bool fieldsInitialized = []() {
+      BrandTypeIterator::initFieldOptions();
+      return true;
+    }();
+  }
+
+  void updateShape(size_t group) {
+    auto [it, inserted] =
+      groupShapeIndices.insert({RecGroupShape(groups[group].group), group});
+    if (inserted) {
+      // This shape was unique. We're done.
+      return;
+    }
```

Comment:
Stupid question maybe, but can't brute force be used here, that is, right after this return we handle *any* conflict with
* advance()
* advanceBrand()
* permute()

It might be a little "wasteful", but if that works it seems simpler?

Reply:
No, that's a good question. It would totally be correct to just do `advance(); permute()` to resolve any conflict. We do need a `GroupClassInfo` in order to call `advance(); permute()`, but we could just initialize every group with a GroupClassInfo eagerly and the disjoint set data structure choose a winner when we combine multiple groups into an equivalence class. We wouldn't even have to do the whole canonicalization thing; we could just choose the permutation of the representative element of each equivalence class as the base permutation for that class.

The problem is that this simpler approach could lead to a large amount of wasted work. First, without our canonicalization routine, there's nothing stopping us from generating a large number of permutations that are automorphic to previous permutations (up to `n! - (n-1)!` of them). Worse, without canonicalization, we might discover that two nontrivial group equivalence classes with different base permutations are actually the same equivalence class, and after combining them, we will start generating permutations using the new representative element that had already been generated by the losing previous representative element. Both of these problems can be solved by advancing the brand whenever we discover an automorphism, but advancing the brand increases code size, so the results are better when we do that as sparingly as possible.

The current approach is a hybrid between the totally lazy approach without any canonicalization described above and another potential approach, which would be to eagerly canonicalize every SCC to sort it directly into its final equivalence class without ever having to "discover" that two types are actually in the same equivalence class. That approach would also be much simpler and would avoid ever generating a useless permutation, but would be far more expensive overall because canonicalization is not cheap.

Reply:
I see, thanks. Makes sense.

How about adding a comment here explaining that we could simplify this that way, at the cost of greater code size, but that if in the future we find that conflicts are rare enough then it may be worthwhile? I'm imagining a future time where we know conflicts are rare + have a reason to refactor this code, and this simplification might make sense then.